# Sprint 10 - Proyecto

La compañía móvil Megaline no está satisfecha al ver que muchos de sus clientes utilizan planes heredados. Quieren desarrollar un modelo que pueda analizar el comportamiento de los clientes y recomendar uno de los nuevos planes de Megaline: Smart o Ultra.

Tienes acceso a los datos de comportamiento de los suscriptores que ya se han cambiado a los planes nuevos (del proyecto del sprint de Análisis estadístico de datos). Para esta tarea de clasificación debes crear un modelo que escoja el plan correcto. Como ya hiciste el paso de procesar los datos, puedes lanzarte directo a crear el modelo.

Desarrolla un modelo con la mayor exactitud posible. En este proyecto, el umbral de exactitud es 0.75. Usa el dataset para comprobar la exactitud.

In [20]:
# Importar las librerías
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

## Paso 1. Abre y examina el archivo de datos

In [2]:
# Leer el dataset
df = pd.read_csv('datasets/users_behavior.csv')

In [3]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB
None


In [4]:
print(df.head())

   calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0


## Paso 2. Segmenta los datos fuente en un conjunto de entrenamiento, uno de validación y uno de prueba.

In [ ]:
# Segmentar los datos
train, other = train_test_split(df, test_size=0.40, random_state=54321)
test, valid = train_test_split(other, test_size=0.50, random_state=54321)

In [9]:
print('Total: ', df.size)
print('Conjunto de entrenamiento', train.size)
print('Conjunto de prueba', test.size)
print('Conjunto de validación', valid.size)

Total:  16070
Conjunto de entrenamiento 9640
Conjunto de prueba 3215
Conjunto de validación 3215


In [10]:
X_train = train.drop('is_ultra', axis=1)
y_train = train['is_ultra']
X_valid = valid.drop('is_ultra', axis=1)
y_valid = valid['is_ultra']
X_test = test.drop('is_ultra', axis=1)
y_test = test['is_ultra']

Cada conjunto de datos se repartio de la siguiente manera:
- Entrenamiento: 60%
- Validación: 20%
- Prueba: 20%

## Paso 3. Investiga la calidad de diferentes modelos cambiando los hiperparámetros.

Modelos a investigar:
- árbol de decisión
- bosque aleatorio
- regresión logística

### Árbol de decisión

In [ ]:
for depth in range(1, 11):
    model_tree = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model_tree.fit(X_train, y_train)
    predic_tree = model_tree.predict(X_valid)
    score_train = model_tree.score(X_train, y_train)
    score_valid = model_tree.score(X_valid, y_valid)

    print(f'Conjunto de entrenamiento - Depth: {depth} - Accuracy: {score_train}')
    print(f'Conjunto de validación - Depth: {depth} - Accuracy: {score_valid}')

Conjunto de entrenamiento - Depth: 1 - Accuracy: 0.7489626556016598
Conjunto de validación - Depth: 1 - Accuracy: 0.7558320373250389
Conjunto de entrenamiento - Depth: 2 - Accuracy: 0.7785269709543569
Conjunto de validación - Depth: 2 - Accuracy: 0.7947122861586314
Conjunto de entrenamiento - Depth: 3 - Accuracy: 0.7946058091286307
Conjunto de validación - Depth: 3 - Accuracy: 0.8040435458786936
Conjunto de entrenamiento - Depth: 4 - Accuracy: 0.7987551867219918
Conjunto de validación - Depth: 4 - Accuracy: 0.7993779160186625
Conjunto de entrenamiento - Depth: 5 - Accuracy: 0.8127593360995851
Conjunto de validación - Depth: 5 - Accuracy: 0.7791601866251944
Conjunto de entrenamiento - Depth: 6 - Accuracy: 0.828838174273859
Conjunto de validación - Depth: 6 - Accuracy: 0.7916018662519441
Conjunto de entrenamiento - Depth: 7 - Accuracy: 0.841804979253112
Conjunto de validación - Depth: 7 - Accuracy: 0.7884914463452566
Conjunto de entrenamiento - Depth: 8 - Accuracy: 0.8589211618257261
Con

### Bosque aleatorio

In [27]:
best_score = 0
best_est = 0
best_depth = 0

for est in range(10, 51, 10):
    for depth in range(1, 11):
        model_forest = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth)
        model_forest.fit(X_train, y_train)
        score_valid = model_forest.score(X_valid, y_valid)
        score_train = model_forest.score(X_train, y_train)

        print('Conjunto de entrenamiento (n_estimators = {}, depth = {}): {})'.format(est, depth, score_train))
        print('Conjunto de validación (n_estimators = {}, depth = {}): {})'.format(est, depth, score_valid))


        if score_valid > best_score:
            best_score = score_valid
            best_est = est
            best_depth = depth

print()
print('La mejor exactitud en el conjunto de validación (n_estimators = {}, depth = {}): {})'.format(best_est, best_depth, best_score))

Conjunto de entrenamiento (n_estimators = 10, depth = 1): 0.7437759336099585)
Conjunto de validación (n_estimators = 10, depth = 1): 0.7465007776049767)
Conjunto de entrenamiento (n_estimators = 10, depth = 2): 0.7728215767634855)
Conjunto de validación (n_estimators = 10, depth = 2): 0.7838258164852255)
Conjunto de entrenamiento (n_estimators = 10, depth = 3): 0.7977178423236515)
Conjunto de validación (n_estimators = 10, depth = 3): 0.7978227060653188)
Conjunto de entrenamiento (n_estimators = 10, depth = 4): 0.8070539419087137)
Conjunto de validación (n_estimators = 10, depth = 4): 0.8102643856920684)
Conjunto de entrenamiento (n_estimators = 10, depth = 5): 0.821058091286307)
Conjunto de validación (n_estimators = 10, depth = 5): 0.8055987558320373)
Conjunto de entrenamiento (n_estimators = 10, depth = 6): 0.8366182572614108)
Conjunto de validación (n_estimators = 10, depth = 6): 0.8055987558320373)
Conjunto de entrenamiento (n_estimators = 10, depth = 7): 0.850103734439834)
Conjun

### Regresión logística

In [21]:
model_logistic = LogisticRegression(random_state=12345, solver='liblinear')
model_logistic.fit(X_train, y_train)
score_train = model_logistic.score(X_train, y_train)
score_valid = model_logistic.score(X_valid, y_valid)

print('Accuracy del modelo de regresión logística en el conjunto de entrenamiento: ', score_train)
print('Accuracy del modelo de regresión logística en el conjunto de validación: ', score_valid)

Accuracy del modelo de regresión logística en el conjunto de entrenamiento:  0.7121369294605809
Accuracy del modelo de regresión logística en el conjunto de validación:  0.7076205287713841


- Se realizó el análisis de los modelos: árbol de decisión, bosque aleatorio, regresión logística, cambiando sus diferentes hiperparámetros. 
- Con la ayuda de bucles se probaron los diferentes hiperparámetros de los 3 diferentes modelos.
- Se imprimió el accuracy con el conjunto de validación y el conjunto de entrenamiento para asegurarnos de que no exista overfitting o underfitting.
- Se encontro la mejor exactitud en el modelo bosque aleatorio con n_estimators = 20 y depth = 5, resultando en el conjunto de validación la puntuación de: 0.8133748055987559.

## Paso 4. Comprueba la calidad del modelo usando el conjunto de prueba

In [28]:
model_forest = RandomForestClassifier(random_state=12345, n_estimators=20, max_depth=5)
model_forest.fit(X_train, y_train)
score = model_forest.score(X_test, y_test)

print('Conjunto de prueba (n_estimators = {}, depth = {}): {})'.format(20, 5, score))

Conjunto de prueba (n_estimators = 20, depth = 5): 0.8102643856920684)


La puntuación máxima de exatitud se alcanzo con el modelo bosque aleatorio con los hiperparámetros n_estimators=20 y depth=5, superando el umbral de 0.75. Dando una exactitud en el conjunto de prueba de 0.81.
